File: C#/dnd-character/DndCharacter.cs
using System;

public class DndCharacter
{
    public int Strength { get; private set; }
    public int Dexterity { get; private set; }
    public int Constitution { get; private set; }
    public int Intelligence { get; private set; }
    public int Wisdom { get; private set; }
    public int Charisma { get; private set; }
    public int Hitpoints { get; private set; }

    public static int Modifier(int score)
    {
        return (int)Math.Floor((score - 10.0) / 2);
    }

    public static int Ability() 
    {
        Random rand = new Random();
        int die1 = rand.Next(1, 7);
        int die2 = rand.Next(1, 7);
        int die3 = rand.Next(1, 7);
        int die4 = rand.Next(1, 7);
        int min1 = Math.Min(die1, die2);
        int min2 = Math.Min(die3, die4);
        int minimum = Math.Min(min1, min2);

        return die1 + die2 + die3 + die4 - minimum;
    }

    public static DndCharacter Generate()
    {
        DndCharacter dnd = new DndCharacter();
        dnd.Strength = Ability();
        dnd.Dexterity = Ability();
        dnd.Constitution = Ability();
        dnd.Intelligence = Ability();
        dnd.Wisdom = Ability();
        dnd.Charisma = Ability();
        dnd.Hitpoints = 10 + Modifier(dnd.Constitution);

        return dnd;
    }
}

File: JavaScript/dnd-character.js
export const abilityModifier = (value) => {
  if (value < 3) {
    throw new Error('Ability scores must be at least 3');
  }
  if (value > 18) {
    throw new Error('Ability scores can be at most 18');
  }
  return Math.floor((value - 10) / 2);
};

export class Character {

  constructor() {
    this.setStrength = Character.rollAbility();
    this.setDexterity = Character.rollAbility();
    this.setConstitution = Character.rollAbility();
    this.setIntelligence = Character.rollAbility();
    this.setWisdom = Character.rollAbility();
    this.setCharisma = Character.rollAbility();
    this.setHitpoints = abilityModifier(this.constitution) + 10;
  }

  static rollAbility() {
    function rollDie() {
      return Math.floor(Math.random() * 6 + 1);
    }

    // create array with 4 dice and roll each one. find the minimum and return the sum - minimum
    let diceArr = [rollDie(), rollDie(), rollDie(), rollDie()];
    let minimum = Math.min(...diceArr);
    return diceArr.reduce((acc, curr) => acc + curr) - minimum;
  }

  get strength() {
    return this.setStrength;
  }

  get dexterity() {
    return this.setDexterity;
  }

  get constitution() {
    return this.setConstitution;
  }

  get intelligence() {
    return this.setIntelligence;
  }

  get wisdom() {
    return this.setWisdom;
  }

  get charisma() {
    return this.setCharisma;
  }

  get hitpoints() {
    return this.setHitpoints;
  }
}
